pkg/runner/cluster_k8s.go

```diff
@@ -0,0 +1,220 @@
+
+	// Build a runenv.
+	runenv := &runtime.RunEnv{
+		TestPlan:           input.TestPlan.Name,
+		TestCase:           testcase.Name,
+		TestRun:            input.RunID,
+		TestCaseSeq:        seq,
+		TestInstanceCount:  input.Instances,
+		TestInstanceParams: input.Parameters,
+		TestSidecar:        true,
+	}
+
+	// Serialize the runenv into env variables to pass to docker.
+	//env := util.ToOptionsSlice(runenv.ToEnvVars())
+
+	env := util.ToEnvVar(runenv.ToEnvVars())
+
+	// Define k8s client configuration
+	config := defaultKubernetesConfig()
+	k8scfg, err := clientcmd.BuildConfigFromFlags("", config.KubeConfigPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not start k8s client from config: %v", err)
+	}
+
+	// Create the clientset
+	clientset, err := kubernetes.NewForConfig(k8scfg)
+	if err != nil {
+		return nil, fmt.Errorf("could not create clientset: %v", err)
+	}
+
+	var (
+		sname    = fmt.Sprintf("tg-%s-%s-%s", input.TestPlan.Name, testcase.Name, input.RunID)
+		replicas = uint64(input.Instances)
+	)
+
+	log.Infow("creating k8s deployment", "name", sname, "image", image, "replicas", replicas)
+
+	jobName := fmt.Sprintf("tg-%s", input.TestPlan.Name)
+
+	// Create Kubernetes Job
+	jobRequest := &v1batch.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: jobName,
+			Labels: map[string]string{
+				"testground.plan":     input.TestPlan.Name,
+				"testground.testcase": testcase.Name,
+				"testground.runid":    input.RunID,
+			},
+		},
+		Spec: v1batch.JobSpec{
+			Parallelism:             int32Ptr(int32(replicas)),
+			Completions:             int32Ptr(int32(replicas)),
+			TTLSecondsAfterFinished: int32Ptr(600),
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"testground.plan":     input.TestPlan.Name,
+						"testground.testcase": testcase.Name,
+						"testground.runid":    input.RunID,
+					},
+				},
+				Spec: v1.PodSpec{
+					RestartPolicy: v1.RestartPolicyNever,
+					Containers: []v1.Container{
+						{
+							Name:  jobName,
+							Image: image,
+							Args:  []string{},
+							Env:   env,
+							Resources: v1.ResourceRequirements{
+								Limits: v1.ResourceList{
+									v1.ResourceMemory: resource.MustParse("30Mi"),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	_, err = clientset.BatchV1().Jobs(config.Namespace).Create(jobRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// Wait for job
+	start := time.Now()
+	for {
+		job, err := clientset.BatchV1().Jobs(config.Namespace).Get(jobName, metav1.GetOptions{})
+		if err != nil {
+			log.Warnw("transient job error", "job", jobName, "err", err)
+			time.Sleep(300 * time.Millisecond)
+			continue
+		}
+
+		log.Debugw("job status", "job", jobName, "active", job.Status.Active, "succeeded", job.Status.Succeeded, "failed", job.Status.Failed)
+		if job.Status.Active == 0 && (job.Status.Succeeded > 0 || job.Status.Failed > 0) {
+			break
+		}
+		if time.Since(start) > 5*time.Minute {
```

Comment:
This will need changing and is worth having a discussion on it - we have to be careful to not `timeout` any plans, but the `timeout` parameter is dependent on scale - 100 pods can be scheduled in a few seconds, 10k pods will be scheduled in a few minutes.

We don't want to be in a situation where test plans start to error out before the last instances have had a chance to start.